serving/src/main/java/feast/serving/service/OnlineServingService.java

```diff
@@ -64,110 +62,211 @@ public GetFeastServingInfoResponse getFeastServingInfo(
+                        return RefUtil.generateFeatureStringRef(featureRef);
+                      },
+                      featureRowField -> {
+                        // drop stale feature values.
+                        if (isStaleValues) {
+                          return Value.newBuilder().build();
+                        } else {
+                          return featureRowField.getValue();
+                        }
+                      }));
+      valueMap.putAll(featureValueMap);
+    }
 
-  private void populateRequestCountMetrics(FeatureSetRequest featureSetRequest) {
-    String project = featureSetRequest.getSpec().getProject();
-    featureSetRequest
-        .getFeatureReferences()
-        .parallelStream()
-        .forEach(ref -> Metrics.requestCount.labels(project, ref.getName()).inc());
-  }
+    // create empty values for features specified in request but not present in feature row.
+    Set<String> missingFeatures =
+        nameRefMap.values().stream()
+            .map(ref -> RefUtil.generateFeatureStringRef(ref))
+            .collect(Collectors.toSet());
+    missingFeatures.removeAll(valueMap.keySet());
+    missingFeatures.forEach(refString -> valueMap.put(refString, Value.newBuilder().build()));
 
-  @Override
-  public GetBatchFeaturesResponse getBatchFeatures(GetBatchFeaturesRequest getFeaturesRequest) {
-    throw Status.UNIMPLEMENTED.withDescription("Method not implemented").asRuntimeException();
+    return valueMap;
   }
 
-  @Override
-  public GetJobResponse getJob(GetJobRequest getJobRequest) {
-    throw Status.UNIMPLEMENTED.withDescription("Method not implemented").asRuntimeException();
+  /**
+   * Generate Field level Status metadata for the given valueMap.
+   *
+   * @param isNotFound whether the given valueMap represents values that were not found in the
+   *     online retriever.
+   * @param isStaleValues whether the given valueMap contains stale values.
+   * @return a 1:1 map containing field status metadata instead of values in the valueMap.
+   */
+  private Map<String, FieldStatus> getMetadataMap(
+      Map<String, Value> valueMap, boolean isNotFound, boolean isStaleValues) {
+    return valueMap.entrySet().stream()
+        .collect(
+            Collectors.toMap(
+                es -> es.getKey(),
+                es -> {
+                  Value fieldValue = es.getValue();
+                  if (isNotFound) {
+                    return FieldStatus.NOT_FOUND;
+                  } else if (isStaleValues) {
+                    return FieldStatus.OUTSIDE_MAX_AGE;
+                  } else if (fieldValue.getValCase().equals(Value.ValCase.VAL_NOT_SET)) {
+                    return FieldStatus.NULL_VALUE;
+                  }
+                  return FieldStatus.PRESENT;
+                }));
   }
 
+  /**
+   * Determine if the feature data in the given feature row is considered stale. Data is considered
```

Comment:
We should really stop using the term stale because it is too overloaded. `isOutsideMaxAge`?

Reply:
Updated.